### ISSUE TITLE
I2S SPH0645LM4H-B problems

### ISSUE THREAD

**Issue by sreimchen** [earlephilhower/arduino-pico#1353]
Hello Earle,

I'm currently using your Arduino core for my master thesis. Thank you for your great work of this project which helps a lot!

Well, sadly I got a problem actually using I2S and I am going out of ideas. I gonna explain step by step what I want to achieve and why it isn't working as I am expecting, maybe you got some time to look over it:

So basically I got a specific sender frequency (here constant 100 Hz). and every 10 ms I want to send all sensor data that is stored via DMA. I looked into your code of the AudioBufferManager class and figured out that I need to configure the buffer size so that everytime a buffer runs full the defined receiver callback is fired. Since I am also using the ADC which is also built on top of the AudioBufferManager I can say that it is working fine in case of the ADC but not in case of the I2S interface (both use the same logic of defining the buffer size that is needed to trigger the interrupt every 10 ms). The ADC is firing the defined callback every 10 ms which I am using internally to trigger a read (and to send the data). That is what I am expecting. I2S instead is firing far slower and also depends anyhow on the set sensor frequency anyhow: 10 kHz ~40 ms, 20 Khz ~70 ms, 30 Khz ~102 ms, 40 Khz ~133 ms (instead of the 10 ms - I am still expecting 10 ms because of the 100 Hz sender frequency that I am configuring the buffer size on). Looks like a factor related to the frequency but I cannot determine what the cause is...

I am going to attach my I2S and ADC files here for further investigation. They might help... maybe I missunderstand the I2S documentation or the protocol itself. Or maybe the sensors, I am using two SPH0645LM4H-B from Adafruit, are causing the problems because I read that these sensors are not using the standard I2S protocol instead they use some different timings. Still and since I get graphs that look fine to me I think that these basically work and this just might cause additional problems but that might be another issue. These are the graphs I get for both mics running in stereo tracked with TelePlot with 20 Khz and 32 bits per sample (because 24 bit caused problems, expecially because the samples are left aligned and the I2S lib of this core is expecting those to be right aligned if I understand correctly. That would cause an unwanted shift to the left which would destroy data):

![image](https://user-images.githubusercontent.com/33686223/229760392-fa25624e-c0ee-43e4-84aa-d7923e54bf16.png)

In case of the 20 kHz and the 32 bits per sample I calculate the buffer size like this:
1. 32 bit per sample correspond to 32 bit buffer words (instead of multiple samples per Word in case of 16 or even 8 bit samples). So one sample (left or right) is exactly one buffer word.
2. This is the formula I am using: bufferSize = wordsPerSample * frequency / Configuration::mqttFrequency * numberOfChannels;
3. So wordsPerSample = 1 as explained in 1., frequency = 20000.0, numberOfChannels = 2 (stereo = 2, mono = 1) because we need to store double the information 
4. Configuration::mqttFrequency is the sender frequency equals to 100 Hz 
5. So in this case: bufferSize = 1 * 20000 / 100 * 2 = 400
6. 400 samples can be stored, 200 for every channel, then is an interrupt (the receiver callback) being fired to read out the exact 400 number of samples, 200 reads because the read function takes both left and right channel paramters

I have not much experience with audio and expecially with I2S. So sorry if I am missing something crucial...

To the files and classes I am attaching: These are wrappers around your classes that implement  just another interface that I am working with but that should not make any difference. I think the calculation of the buffer size and the configuration of the interfaces itself might be the cause... 

Here are some interesting sources to the specific I2S mic that I use (SPH0645LM4H-B):
Datasheet: https://cdn-shop.adafruit.com/product-files/3421/i2S+Datasheet.PDF
Difference to other I2S mics: https://forum.pjrc.com/threads/47010-I2S-Microphone-(SPH0645LM4H-B)?highlight=microphone
Youtube video to this exact mic: https://www.youtube.com/watch?v=3g7l5bm7fZ8&ab_channel=atomic14
Mixed left and right channel of this mic: https://www.atomic14.com/2020/09/12/esp32-audio-input.html

If something is not clear, please tell me... thanks a lot for your time!

C++ files:
[cpp-files.zip](https://github.com/earlephilhower/arduino-pico/files/11147597/cpp-files.zip)

**Comment by earlephilhower** [comment]
You've got a ton of information, but can you summarize exactly what you expect and aren't getting, along with enough code to reproduce the failure (i.e. MCVE, not your full app....make it fit in 20 lines of `setup()` or something so it's parseable w/o knowing your whole app or data flow)?  Without that it's almost impossible to help you.

Some general notes from skimming through your posts:

`I2S::setFrequency` takes the word clock frequency, no need to multiply by 2.  If you expect 44.1K samples from left and 44.1K samples from the right then `setFrequency(44100)` (freq obviously implies a high and low period on the clock...or it would just be DC).

You can't change any configuration after I2S::begin() so make sure you've set everything as you like it before starting.

Also, there is only LSBJ output support, not input.  Never seen a Japanese-format mike and even the DACs are pretty rare.  So no need to setLSBJFormat(), not sure why you posted that.

The I2S buffer size is in 32B words, but with 32b samples then the buffer will have 1/2 the left/right pairs as the # you pass in.  There will be some jitter due to IRQ timing, etc, but over large enough samples you should be getting DMA callbacks at the appropriate period.

There's also no need to have anything connected or look at mike output waveforms....the I2S protocol is fire-and-forget so the Pico will send out clocks and sample DIN at the appropriate frequency whether there's anything connected or not.

**Comment by sreimchen** [comment]
Hi, sorry, for my late response and large partially unrelated post. And thank you for your quick and helpful answer!

I broke the problem down and saw that the problem lies somewhere else: The tranmission via Serial seems to be the bottleneck that delays outgoing messages (of course... after calculating an estimate of the message sizes and their rate that makes sense...). The higher frequencies I used the higher the delay was. Without transmitting any raw data and instead printing just the intervals between the IRQ's it does seem to work. In my main project it already looked like this specific microphone is also working (just not with higher frequencies because I was sending the raw data which ran fast in the explained bottleneck). 

Sorry, for taking your time... I should have tried breaking it down by myself right at the beginning before posting here. 

I wish happy easter days!
